src/Libraries/Microsoft.Extensions.AI/ChatCompletion/CachingChatClient.cs

```diff
@@ -50,58 +67,126 @@ public override async IAsyncEnumerable<StreamingChatCompletionUpdate> CompleteSt
-#pragma warning restore S1643
-                }
-                else
-                {
-                    capturedItems.Add(item);
+                    // If we couldn't find anything to coalesce, there's nothing to do.
+                    if (endExclusive - startInclusive <= 1)
+                    {
+                        continue;
+                    }
+
+                    // We found a coalescable run of items. Create a new node to represent the run. We create a new one
+                    // rather than reappropriating one of the existing ones so as not to mutate an item already yielded.
+                    _ = coalescedText.Clear().Append(capturedItems[startInclusive].Text);
+
+                    TextContent coalescedContent = new(null) // will patch the text after examining all items in the run
+                    {
+                        AdditionalProperties = textContent.AdditionalProperties?.Clone(),
+                        ModelId = textContent.ModelId,
+                    };
+
+                    StreamingChatCompletionUpdate coalesced = new()
+                    {
+                        AdditionalProperties = update.AdditionalProperties?.Clone(),
+                        AuthorName = update.AuthorName,
+                        CompletionId = update.CompletionId,
+                        Contents = [coalescedContent],
+                        CreatedAt = update.CreatedAt,
+                        FinishReason = update.FinishReason,
+                        Role = update.Role,
+
+                        // Explicitly don't include RawRepresentation. It's not applicable if one update ends up being used
+                        // to represent multiple, and it won't be serialized anyway.
+                    };
+
+                    // Replace the starting node with the coalesced node.
+                    capturedItems[startInclusive] = coalesced;
+
+                    // Now iterate through all the rest of the updates in the run, updating the coalesced node with relevant properties,
+                    // and nulling out the nodes along the way. We do this rather than removing the entry in order to avoid an O(N^2) operation.
+                    // We'll remove all the null entries at the end of the loop, using RemoveAll to do so, which can remove all of
+                    // the nulls in a single O(N) pass.
+                    for (int i = startInclusive + 1; i < endExclusive; i++)
+                    {
+                        // Grab the next item.
+                        StreamingChatCompletionUpdate next = capturedItems[i];
+                        capturedItems[i] = null!;
+
+                        TextContent nextContent = (TextContent)next.Contents[0];
+                        _ = coalescedText.Append(nextContent.Text);
+
+                        coalesced.AdditionalProperties = MergeProperties(coalesced.AdditionalProperties, next.AdditionalProperties);
```

Comment:
This seems potentially very expensive. Streaming output tends to be chunked on a very fine-grained level, usually individual words or maybe even tokens. Performing a full dictionary merge for every chunk could be a lot of work.

I know it's still linear in the response length, assuming the number of additional properties per chunk is unrelated to the number of chunks. But still potentially a lot.

Do you think there are definitely cases where a developer would expect to see a merged dictionary? And why would it be correct to have a "last chunk wins" rule for properties that vary on a per-chunk basis? Would be reasonable to say we just take metadata from the first chunk, and if a developer wants to preserve full granularity, they can disable the coalescing?

Reply:
> Do you think there are definitely cases where a developer would expect to see a merged dictionary?

There are cases where properties show up at some point after not having been on previous chunks. I'm not particularly concerned with cases where the same property shows up multiple times or with the last-write-wins policy, it was just easiest to implement, with the indexer (TryAdd would be first-wins but doesn't exist on netfx, and if we're concerned about cost, implementing it with TryGetValue and TryAdd makes it more exoensive).

I don't feel super strong about it, though. If you don't think it's worth it, I can remove the dictionary stuff.